src/runtime/nativecall.cs

```diff
@@ -40,10 +40,15 @@ public static int Int_Call_3(IntPtr fp, IntPtr a1, IntPtr a2, IntPtr a3)
             return d(a1, a2, a3);
         }
 
-        private static T GetDelegate<T>(IntPtr fp) where T: Delegate
+        internal static T GetDelegate<T>(IntPtr fp) where T: Delegate
         {
-            // Use Marshal.GetDelegateForFunctionPointer<> directly after upgrade the framework
-            return (T)Marshal.GetDelegateForFunctionPointer(fp, typeof(T));
+            Delegate d = null;
+            if (!Interop.allocatedThunks.TryGetValue(fp, out d))
+            {
+                // Use Marshal.GetDelegateForFunctionPointer<> directly after upgrade the framework
+                d = Marshal.GetDelegateForFunctionPointer(fp, typeof(T));
```

Comment:
Can you also apply the comment here? The generic function was introduced in .NET 4.5.1.

Reply:
Actually, maybe just remove the comment. You are not writing this into `allocatedThunks` right now, btw.

Reply:
Yes, I don't cache these because they truly point to unmanaged code. We should only cache the function pointers to managed code.